### DISCUSSION TITLE
USB descriptors help!

### DISCUSSION THREAD

**Discussion by DebugerDoesntWork** [hathach/tinyusb#1955] [Q&A]
Hey there! I am trying to use a TinyUSB for the first time and I am finding it extremely difficult to get my USB descriptor files right. This is my first time programming USB anything ever and I'm not sure how to get this right. About the project: I am making a custom keyboard as a fun project. To be honest, I tried playing around with it for over a day and I couldn't get it to work. I started with the base descriptor file provided in some example folders and spent a lot of time trying to modify it with no success. I am using the raspberry pi pico w as my board, but I'm not using it wirelessly. I also don't have a vendor ID or product ID as I am just a student, I heard that some companies are willing to allow you to use their VIDs but I wasn't able to find any information about this. Has anyone used something like that before? As of right now, I've been compiling with the default 0xCafe VID, could that be causing the issue?
I have tried to look for help in the discussions and ChatGPT/Bing chat for help but I still can't figure it out. Thank you so much for your help! Also, thank you so much for this awesome project!

**Comment by hathach** [comment]
too long code hurt readibility, please edit your post to attach an txt file instead.

**Reply by DebugerDoesntWork** [reply]
Sorry about that, I've never used GitHub before. Here's the code.

[usbdesc.txt](https://github.com/hathach/tinyusb/files/11069780/usbdesc.txt)

**Reply by hathach** [reply]
you also need to remove the lenghty code in your 1st post

**Reply by DebugerDoesntWork** [reply]
> you also need to remove the lenghty code in your 1st post

Done!

**Reply by hathach** [reply]
I haven't looked at your code yet, but there is a few way we could analys it first
- which OS you are using, if linux check your dmesg, windows check the device manager to see if there is error in your descriptors. 
- You can also check your descriptor with online parser https://eleccelerator.com/usbdescreqparser/
- compile LOG=2 (CFG_TUSB_DEUBG=2) and get debug log from rp2040 UART, it may contain hint/error 

> I also don't have a vendor ID or product ID as I am just a student, I heard that some companies are willing to allow you to use their VIDs but I wasn't able to find any information about this. Has anyone used something like that before? As of right now, I've been compiling with the default 0xCafe VID, could that be causing the issue?

If you don't plan to sell your own product comercially, it is ok to use 0xcafe or any VID you pick, since it is only attached to handful of PCs. Note: it does violate the VID usage by USB-IF though.

**Reply by DebugerDoesntWork** [reply]
Thank you so much for the tips!
